hkmc2/shared/src/main/scala/hkmc2/bbml/bbML.scala

```diff
@@ -57,27 +67,38 @@ object BbCtx:
   def refTy(ct: Type, sk: Type)(using ctx: BbCtx): Type =
     ClassLikeType(ctx.getCls("Ref").get, Wildcard(ct, ct) :: Wildcard.out(sk) :: Nil)
   def init(raise: Raise)(using Elaborator.State, Elaborator.Ctx): BbCtx =
-    new BbCtx(raise, summon, None, 1, HashMap.empty)
+    new BbCtx(raise, summon, None, 1, HashMap.empty, Bot, N)
 
   val builtinOps = Set("+", "-", "*", "/", "<", ">", "<=", ">=", "==", "!=", "&&", "||")
 end BbCtx
 
 
-class BBTyper(using elState: Elaborator.State, tl: TL, scope: Scope):
+class BBTyper(using elState: Elaborator.State, tl: TL):
   import tl.{trace, log}
   
   private val infVarState = new InfVarUid.State()
-  private val solver = new ConstraintSolver(infVarState, tl)
+  private val solver = new ConstraintSolver(infVarState, elState, tl)
 
   private def freshSkolem(sym: Symbol, hint: Str = "")(using ctx: BbCtx): InfVar =
     InfVar(ctx.lvl, infVarState.nextUid, new VarState(), true)(sym, hint)
   private def freshVar(sym: Symbol, hint: Str = "")(using ctx: BbCtx): InfVar =
     InfVar(ctx.lvl, infVarState.nextUid, new VarState(), false)(sym, hint)
   private def freshWildcard(sym: Symbol)(using ctx: BbCtx) =
-    val in = freshVar(sym)
-    val out = freshVar(sym)
+    val in = freshVar(sym, "-")
+    val out = freshVar(sym, "+")
     // in.state.upperBounds ::= out // * Not needed for soundness; complicates inferred types
     Wildcard(in, out)
+  private def freshReg(sym: Symbol)(using ctx: BbCtx) =
+    val state = new VarState()
+    state.upperBounds = ctx.getRegEnv.! :: Nil
+    InfVar(ctx.lvl + 1, infVarState.nextUid, state, true)(sym, "")
+  private def freshOuter(sym: Symbol)(using ctx: BbCtx): InfVar =
+    InfVar(ctx.lvl + 1, infVarState.nextUid, new VarState(), true)(sym, "")
+  private def freshEnv(sym: Symbol)(using ctx: BbCtx): InfVar =
+    val state = new VarState()
+    state.upperBounds = ctx.getRegEnv :: Nil
+    state.lowerBounds = ctx.getRegEnv :: Nil
```

Comment:
I wonder what's the point of creating a new TV that's equivalent to an underlying type `ctx.getRegEnv`. I see `freshEnv` is only used in `instantiate`. Why can't `instantiate` substitute simple types for variables? Instead of just variables for variables?